STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -88,72 +166,83 @@ uint32_t computeCRC(CRC_HandleTypeDef *hcrc, uint8_t *data, uint32_t size)
+    
+    // User is trying to write to an address containing program data
+    if (isProgramMemory(flash_address, size)) { return 1; }
+
     uint32_t flash_sector = getFlashSector(flash_address);
 
-    // Erase the sector before write
-    HAL_FLASH_Unlock();
+    // User is trying to write to non-valid sector
+    if (flash_sector == 10) { return 1; }
+
+    // User is trying to write to more than one sector
+    if (isWriteWithinSector(flash_address, size) != 0){ return 1; }
+
+    // If !HAL_OK FLASH unlock failed and cannot be erased or written to
+    if (HAL_FLASH_Unlock() != HAL_OK) { return 1; }
+
+    // Erase the sector before write.
     FLASH_Erase_Sector(flash_sector, FLASH_VOLTAGE_RANGE_3);
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
 
     // Compute CRC of data to be saved.
     uint32_t crcVal = computeCRC(hcrc, data, size);
 
-    // Append CRC converted to uint8_t to data
-    uint8_t crcData[4] = {0};
-    for (uint8_t i = 0; i < 4; i++)
-    {
-        uint8_t convertedCrcTemp = (crcVal >> 8*i) & 0xFF;
-        //*(data + size + i) = convertedCrcTemp;
-        crcData[i] = convertedCrcTemp;
-    }
-
     __HAL_RCC_WWDG_CLK_DISABLE();
-    // Write to sector
     HAL_FLASH_Unlock();
-    // Save data supplied as input
+    // Store data
     for(uint32_t i=0; i<size; i++)
     {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]);
+        if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]) != HAL_OK) { return 1; }
     }
 
-    // Save CRC value
-    for(uint32_t i=0; i<4; i++)
-    {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+size+i , crcData[i]);
-    }
+    // Store CRC value
+    if (HAL_FLASH_Program(FLASH_TYPEPROGRAM_WORD, flash_address+size , crcVal) != HAL_OK) { return 1; }
+
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
+
+    return 0;
 }
 
 
 /*!
 ** @brief Read data from FLASH memory at flash_adress and verify data with stored CRC.
+** @return Returns 1 on error and 0 on success. *data is unchanged on error.
 */
-void readFromFlashCRC(CRC_HandleTypeDef *hcrc, uint32_t flash_address, uint8_t *data, uint32_t size)
+int readFromFlashCRC(CRC_HandleTypeDef *hcrc, uint32_t flash_address, uint8_t *data, uint32_t size)
 {
-    // Read data including CRC value
-    for(uint32_t i=0; i<size; i++)
-    {
-        *(data + i) = *( (uint8_t *)(flash_address+i) );
-    }
+    // User is trying to read from an address containing program data
+    if (isProgramMemory(flash_address, size)) { return 1; }
+
+    uint32_t flash_sector = getFlashSector(flash_address);
+    // User is trying to read from a non-valid sector
+    if (flash_sector == 10) { return 1; }
+
+    // Create temporary storage for FLASH data
+    uint8_t dataTmp[size];
+    memcpy(&dataTmp, (void*) flash_address, size);
 
     // Compute CRC of data read.
-    uint32_t crcVal = computeCRC(hcrc, data, size);
+    uint32_t crcVal = computeCRC(hcrc, dataTmp, size);
```

Comment:
Can you not computeCRC directly on the flash address?

Reply:
The `HAL_CRC_Calculate(...)` takes in data buffer and not a pointer. Hence, we need to retrieve the data first.